executor/slow_query.go

```diff
@@ -808,55 +858,133 @@ func (e *slowQueryRetriever) getFileStartTime(file *os.File) (time.Time, error)
+	if !ok {
+		return
+	}
+	s.totalFileNum += tmp.totalFileNum
+	s.readFileNum += tmp.readFileNum
+	s.readFile += tmp.readFile
+	s.initialize += tmp.initialize
+	s.readFileSize += tmp.readFileSize
+	s.parseLog += tmp.parseLog
+}
+
+// Clone implements the RuntimeStats interface.
+func (s *slowQueryRuntimeStats) Clone() execdetails.RuntimeStats {
+	newRs := *s
+	return &newRs
+}
+
+// Tp implements the RuntimeStats interface.
+func (s *slowQueryRuntimeStats) Tp() int {
+	return execdetails.TpSlowQueryRuntimeStat
+}
+
 func (e *slowQueryRetriever) getFileEndTime(file *os.File) (time.Time, error) {
 	var t time.Time
+	var tried int
 	stat, err := file.Stat()
 	if err != nil {
 		return t, err
 	}
-	fileSize := stat.Size()
-	cursor := int64(0)
-	line := make([]byte, 0, 64)
+	endCursor := stat.Size()
 	maxLineNum := 128
-	tryGetTime := func(line []byte) string {
-		for i, j := 0, len(line)-1; i < j; i, j = i+1, j-1 {
-			line[i], line[j] = line[j], line[i]
+	for {
+		lines, readBytes, err := readLastLines(file, endCursor)
+		if err != nil {
+			return t, err
+		}
+		// read out the file
+		if readBytes == 0 {
+			break
+		}
+		endCursor -= int64(readBytes)
+		for i := len(lines) - 1; i >= 0; i-- {
+			if strings.HasPrefix(lines[i], variable.SlowLogStartPrefixStr) {
+				return ParseTime(lines[i][len(variable.SlowLogStartPrefixStr):])
+			}
 		}
-		lineStr := string(line)
-		lineStr = strings.TrimSpace(lineStr)
-		if strings.HasPrefix(lineStr, variable.SlowLogStartPrefixStr) {
-			return lineStr[len(variable.SlowLogStartPrefixStr):]
+		tried += len(lines)
+		if tried >= maxLineNum {
+			break
 		}
-		return ""
 	}
+	return t, errors.Errorf("invalid slow query file %v", file.Name())
+}
+
+// Read lines from the end of a file
+// endCursor initial value should be the filesize
+func readLastLines(file *os.File, endCursor int64) ([]string, int, error) {
+	var lines []byte
+	var firstNonNewlinePos int
+	var cursor = endCursor
 	for {
-		cursor -= 1
-		_, err := file.Seek(cursor, io.SeekEnd)
-		if err != nil {
-			return t, err
+		// stop if we are at the beginning
+		// check it in the start to avoid read beyond the size
+		if cursor <= 0 {
+			break
+		}
+
+		var size int64 = 4096
+		if cursor < size {
+			size = cursor
 		}
+		cursor -= size
 
-		char := make([]byte, 1)
-		_, err = file.Read(char)
+		_, err := file.Seek(cursor, io.SeekStart)
 		if err != nil {
-			return t, err
+			return nil, 0, err
 		}
-		// If find a line.
-		if cursor != -1 && (char[0] == '\n' || char[0] == '\r') {
-			if timeStr := tryGetTime(line); len(timeStr) > 0 {
-				return ParseTime(timeStr)
-			}
-			line = line[:0]
-			maxLineNum -= 1
+		chars := make([]byte, size)
+		_, err = file.Read(chars)
+		if err != nil {
+			return nil, 0, err
 		}
-		line = append(line, char[0])
-		if cursor == -fileSize || maxLineNum <= 0 {
-			if timeStr := tryGetTime(line); len(timeStr) > 0 {
-				return ParseTime(timeStr)
+		lines = append(chars, lines...)
```

Comment:
why reverse the append's order?
```
... [block3] [block2] [block1]?
```

Reply:
Yes, this function needs to read the last line first. 
actually this function is port from https://github.com/pingcap/sysutil/pull/19/files#r451300500

Reply:
good, intertesting